### ISSUE TITLE
Public API for Date::added() and Date::until()

### ISSUE THREAD

**Issue by hsivonen** [unicode-org/icu4x#3964]
Temporal requires https://tc39.es/proposal-temporal/#sec-temporal-calendardateaddition and https://tc39.es/proposal-temporal/#sec-temporal-calendardatedifference , but these are unstable hidden APIs in ICU4X. For Temporal compat, we need public API surface for these.

**Comment by sffc** [comment]
Given that we currently have experimental code for this, and it ties in a lot with Temporal and the calendar work we've been doing, I think it should be a 1.x priority.

**Comment by sffc** [comment]
https://tc39.es/proposal-temporal/#sec-temporal-calendardateadd

**Comment by sffc** [comment]
Also see https://github.com/tc39/proposal-intl-era-monthcode/issues/32

**Comment by Manishearth** [comment]
We have these APIs but they are currently unstable. TO be graduated in 2.2

**Comment by Manishearth** [comment]
See also: https://github.com/unicode-org/icu4x/issues/7161

**Comment by sffc** [comment]
Some considerations for naming these fns:

1. Temporal has a `Reject` option in `add`, so we need fallible versions of `add`.
2. However, the `Constrain` option makes `add` infallible, except it can still go out of bounds (#7207), unless we want to make it saturate like `Date::from_rata_die`.
3. `until` is infallible except when the dates are in different calendars. Currently on main we have `try_until_with_options` which you can exhaustively match thanks to an `Infallible` error type on fixed calendars.

Here's a proposed solution:

1. Keep `try_add_with_options` and `try_until_with_options` as today
2. Add `until` gated on the error variant being `Infallible` (it won't appear on `Date<AnyCalendar>`)
3. Make `add` with overflow `Constrain` saturate on out-of-bounds values
4. Add `add` that uses constrain behavior
5. Do not add operators (`core::ops`) at this time, but we could consider the request if a client asks for it

Thoughts?

CC @Manishearth @robertbastian @nekevss

**Comment by Manishearth** [comment]
I generally feel fine with just having the fallible APIs. I am not opposed to infallible convenience APIs, but I don't think we need them, especially not for graduation.

Proposal:

 - Everything is fallible, `try_added_with_options`/`try_until_with_options` and maybe `try_add_with_options` (mutation APIs have subtle issues over FFI)
 - We file an issue for adding infallible variants, and let that resolve itself over a non-release-tied timeline, perhaps waiting for user feedback on our main APIs first.

**Comment by sffc** [comment]
How did you feel about:

> Make `add` with overflow `Constrain` saturate on out-of-bounds values

**Comment by Manishearth** [comment]
I covered that: not opposed to it, but don't think we need it, especially not for graduation.

I'm not really _excited_ about that interpretation of `add`, if that's what you're asking.

**Comment by sffc** [comment]
I agree that the extra APIs are not necessary for graduation, but I think we must decide the behavior of `.add` when the result goes out of bounds. Related: #7207

**Comment by sffc** [comment]
This subject is right in the middle of this issue and #7207 so I'll keep the discussion here since this is where we've been having it so far.

Options for what happens when arithmetic exceeds the range decided in #7207:

1. Always returns an error
2. Returns an error if overflow is Reject, saturates (to the hard RD limit) if overflow is Constrain
3. Always saturates
4. Something else?

**Comment by Manishearth** [comment]
I would like to always return an error. I don't think saturation on _date_ arithmetic is a useful operation.

 `Overflow::Constrain` already has a useful meaning in date arithmetic, e.g. around what you do when you try `2024-02-29 + P1Y`. This overflow parameter is about handling overflows when the arithmetic is producing a _calendrically_ invalid date.

Whereas these ranges are ICU4X-specific things. When someone tries `2024-01-01 + P100000000000Y` there is nothing calendrically incorrect about that operation[^1], we're just saying "ICU4X is not the library for you if you want to do that". I consider that to be a very different kind of constraining.

Furthermore, from a Temporal client perspective, overloading `Constrain` to mean this might require Temporal clients to try and tell the difference between "constrained due to algorithm behavior" vs "constrained due to ICU4X limits". Though I suppose by and large Temporal's ±270k range check would handle that anyway for saturated dates.

I would not be strongly opposed to adding this behavior as a second `date_range_overflow: Option<Overflow>` field to `DateAddOptions`. This lack of opposition can turn into a vote for support if we have users who want this. I am in favor of filing a followup with that proposal, linking to it from the docs, and waiting to see if anyone asks.

 [^1]: Putting aside my usual opinion that it is pointless to even be thinking of dates that far in the future.

**Comment by sffc** [comment]
OK, that's fair. I'm glad we have a path, with `DateAddOptions`, to get the infallible behavior.

**Comment by Manishearth** [comment]
I added a note to file followups to the checklist in the issue. We should do that once we have consensus on what the graduated API will be.

**Comment by sffc** [comment]
Discussion with @sffc @robertbastian:

- @robertbastian If we return an error, what is the error? Is it a range error over years? No, because it could be in the middle of a year. But we don't want to return a range error over RDs because it's not actionable for the user.
- @sffc It's actionable in the sense that their duration is too big.
- @robertbastian Right, so we should return just an Overflow error. I don't want RD range to be something that the client ever needs to worry about.
- @sffc I'm okay using a new `Overflow` error variant on DateAddError.
- @robertbastian Ok. `OverflowOrUnderflow`? I think I would still rather saturate.

**Comment by Manishearth** [comment]
Overflow or OverflowOrUnderflow is fine. IME an underflow is a type of overflow so we don't need to specify.

**Comment by sffc** [comment]
To me "underflow" means "too close to zero", but this is "too close to minus infinity", which to me is "overflow".

**Comment by sffc** [comment]
Agreed to add an Overflow error variant: https://github.com/unicode-org/icu4x/issues/7076#issuecomment-3879635822

**Comment by sffc** [comment]
Updated proposal:

1. Keep `try_add_with_options`, `try_added_with_options`, and `try_until_with_options` in Rust
2. The adding functions should return `DateAddError`
3. One of the error variants will be `DateAddError::Overflow`
4. Export FFI-friendly versions of these to be reviewed by @Manishearth 
5. Do not add `until` at this time. Open an issue in the Priority Backlog milestone to gather feedback. Acknowledge that the function is low cost and can be implemented infallibly on top of `try_until_with_options`.
6. Do not use `overflow: Constrain` as a hint for what to do on RD overflow. Open an issue in the Backlog milestone to collect feedback on whether we should have an infallible `add` function.
7. Do not add operators (`core::ops`) at this time. Open an issue in the Backlog milestone to collect feedback on whether the operators would be handy.

Does this work for you @Manishearth ?

**Comment by Manishearth** [comment]
I like this plan! This gets us the APIs needed by clients which have been baking for one release, without blocking us out of further improvements based on real world needs.

**Comment by robertbastian** [comment]
> * Do not use `overflow: Constrain` as a hint for what to do on RD overflow. Open an issue in the Backlog milestone to collect feedback on whether we should have an infallible `add` function.

What does this mean?

**Comment by Manishearth** [comment]
This means that RD overflow will always return an error, regardless of `overflow`. `overflow: Constrain` remains about  _field_ overflow, e.g. "you produced Feb 30, let me constrain that down to 29 or 28", whereas our internal RD limit cannot be saturated to by most user APIs (`from_rata_die` is a legacy exception).

**Comment by Manishearth** [comment]
I believe the current API matches the plan above. All that is left is graduation review.

I filed the followup backlog issue for `Date::add()` https://github.com/unicode-org/icu4x/issues/7694

**Comment by Manishearth** [comment]
Hmm, `Date::try_added_with_options()` takes `self` not `&self`, which makes some sense since it needs to clone the calendar (which is cheap).

I think that's _fine_ but it's somewhat surprising. We could take `&self` and mandatorily internally clone, but that's a bit of an antipattern.

(cloning all AnyCalendar dates should be cheap anyway)

**Comment by sffc** [comment]
`try_added_with_options` takes by value so it can be used in a chain.

**Comment by robertbastian** [comment]
That's not a requirement for chaining.

**Comment by Manishearth** [comment]
We should check if we want AnyCalendarDifferenceError to be a struct or enum in both Rust and FFI.

Edit: https://github.com/unicode-org/icu4x/pull/7713

**Comment by Manishearth** [comment]
Basically, the tradeoff for `try_added_with_options` is:

If it takes `self`, people need to call `.clone()` on it when they want to retain their original date. This is cheap as long as they are not allocating a Calendar and using builtin calendars. it is semi-expensive if they decide to `Box<Calendar>` or use a custom data loading calendar. In the cases where it is cheap, it is _annoying_ to type. In the cases where it is expensive, being forced to type it may make people choose to use `Ref`/etc.

If it takes `&self`, then it internally calls `.clone()`. This has the same performance characteristics for before, but now it is mandatory for everyone, so people in the "semi expensive" case pay a hidden cost. Not noticing the cost, they may not rearchitect their code to avoid it.

That cost must happen either way, the question sort of is if we make it annoying for the people in the cheap case or if we make it unobvious for the people in the expensive case. I consider the expensive case to be sufficiently niche that I don't mind the unobvious cost.

**Comment by robertbastian** [comment]
We should not let non-Copy calendars drive our API decisions anymore. I'm fine with calling clone internally.

**Comment by Manishearth** [comment]
#7714 for `self`

**Comment by sffc** [comment]
I'm definitely opposed to an implicit clone.

People can avoid the clone by calling `.add` instead. If they are given a reference, then they need to call `.clone()` explicitly with a comment saying "AnyCalendar clones are cheap" or "this uses `Ref<C>` so clones are cheap" or similar. They have no _opportunity to discover_ that they need to do this if we hide away the `.clone()` except for being avid readers of the docs.

Or perhaps we can define `.added(&self)` only when `Ref<C>` is used?

(I'm also opposed to any logic stemming from AnyCalendar planning to contain only Copy calendars in 2.2, but that's not the core issue here.)

**Comment by robertbastian** [comment]
That sounds very much like the core issue. If we made `AnyCalendar` `Copy`, we wouldn't have this discussion.

**Comment by sffc** [comment]
> That sounds very much like the core issue. If we made `AnyCalendar` `Copy`, we wouldn't have this discussion.

Even if AnyCalendar implemented Copy, something I do not currently think we should do, there should still be no implicit cloning inside Date. That's why AnyCalendar is not the "core issue."

**Comment by Manishearth** [comment]
> People can avoid the clone by calling .add instead. 

That assumes they are okay with their date being mutated, which is often not the case. So they'd have to clone there as well; this just moves the clone around.

> explicitly with a comment saying "AnyCalendar clones are cheap" or "this uses Ref<C> so clones are cheap" or similar

Most users are not going to know that AnyCalendar clones are cheap, OR that the rest of Date is cheap to clone[^1]. That's part of the problem here; our API forces a clone in a context where it is non-obvious where it is cheap. This will more likely force users to try hard to avoid the clone, potentially wasting time with `Rc<Date>` (won't work) or something else.

The "avid users of docs" argument cuts both ways: only avid users of docs will know what clones are cheap. Not knowing that a clone is cheap, a Rust user is quite likely to go down the wrong path rather than figuring out if a clone is cheap, in my experience.

I think using `Date` with an expensive-to-clone `AsCalendar` type is a pretty niche use case overall and I am fine making things less discoverable in that case. Niche use case users _should_ be avid readers of docs.

Note that even if AnyCalendar is not Copy, I think any further additions ought to be _cheap_ to clone (i.e. [`Claim`](https://smallcultfollowing.com/babysteps/blog/2024/06/26/claim-followup-1/), if that ever happens) when used on the happy path. E.g. most DataPayload variants are supposed to be cheap clones (at most an Rc clone) unless you are custom data loading.

> Even if AnyCalendar implemented Copy, something I do not currently think we should do, there should still be no implicit cloning inside Date

I found this argument needs expanding, so I shall expand what I perceive it to be: there should be no implicit cloning inside Date because Date contains an `AsCalendar` which may or may not be cheap to clone (`Claim`?).

I agree with that logic. But if AnyCalendar and AnyCalendarDateInner were both Copy, then `Date<AnyCalendar>` would be Copy, and there would be no user ergonomics issue of `.try_added_with_options()` taking `self` by value. That's another way of solving the ergonomics issue here, which makes it relevant to this discussion.

 [^1]: DateInner can be figured out by following it on the trait but it's still non obvious

**Comment by sffc** [comment]
> > People can avoid the clone by calling .add instead.
> 
> That assumes they are okay with their date being mutated, which is often not the case. So they'd have to clone there as well; this just moves the clone around.

I mean, they can either architect their code to use a mutable date, or they can clone with justification. Both are fine.

> The "avid users of docs" argument cuts both ways: only avid users of docs will know what clones are cheap. Not knowing that a clone is cheap, a Rust user is quite likely to go down the wrong path rather than figuring out if a clone is cheap, in my experience.

If there is a `.clone()` in code, it instigates the reading of the docs. It's like if I have to write the word `unstable`, `unsafe`, `unwrap`: it makes me read the docs to double-check that what I'm doing is right, and also helps the reviewer verify it.

> > Even if AnyCalendar implemented Copy, something I do not currently think we should do, there should still be no implicit cloning inside Date
> 
> I found this argument needs expanding, so I shall expand what I perceive it to be: there should be no implicit cloning inside Date because Date contains an `AsCalendar` which may or may not be cheap to clone (`Claim`?).

Right, we designed `Date` early on to avoid cloning calendars. It's why we have `Ref` and `AsCalendar`. People have been landing code to make more of the built-in calendars be `Copy` (cool). I don't see that as having any impact though on the design of `Date` treating the calendar as something not to be implicitly cloned.

>  if AnyCalendar and AnyCalendarDateInner were both Copy, then `Date<AnyCalendar>` would be Copy, and there would be no user ergonomics issue of `.try_added_with_options()` taking `self` by value.

Sure, and in that world, `.try_added_with_options()` should take `self` by value. So we should make it take `self` by value.

**Comment by Manishearth** [comment]
> I mean, they can either architect their code to use a mutable date, or they can clone with justification. Both are fine.

Yes? I have at no point said that this is prohibitive. I am saying that this is annoyance on the happy path.

> If there is a .clone() in code, it instigates the reading of the docs. It's like if I have to write the word unstable, unsafe, unwrap: it makes me read the docs to double-check that what I'm doing is right, and also helps the reviewer verify it.


I do not think the cheap copies here are obvious, and I'm not sure the best way to make that clear. That is my worry: these are trickier generic types: you have to juggle DateInner, AnyDateInner, AsCalendar, etc to figure out if everything is a cheap copy.

> Right, we designed Date early on to avoid cloning calendars. It's why we have Ref and AsCalendar. People have been landing code to make more of the built-in calendars be Copy (cool). I don't see that as having any impact though on the design of Date treating the calendar as something not to be implicitly cloned.



Okay, that's a fair perspective.

I'm okay with reverting the change, but I would like that to be paired with proper documentation that makes it an easy determination for when `Date` is a cheap copy.

**Comment by robertbastian** [comment]
> > if AnyCalendar and AnyCalendarDateInner were both Copy, then `Date<AnyCalendar>` would be Copy, and there would be no user ergonomics issue of `.try_added_with_options()` taking `self` by value.

Just for the record, `AnyDateInner` is `Copy`, `Calendar::DateInner` has a `Copy` bound.

> Sure, and in that world, `.try_added_with_options()` should take `self` by value. So we should make it take `self` by value.

> Right, we designed `Date` early on to avoid cloning calendars. It's why we have `Ref` and `AsCalendar`. People have been landing code to make more of the built-in calendars be `Copy` (cool). I don't see that as having any impact though on the design of `Date` treating the calendar as something not to be implicitly cloned.

And unfortunately we stopped short of making `AnyCalendar` `Copy`. The combination of having what is probably our main calendar type be non-`Copy`, and these APIs taking `Date`s by value, means people will start allocating and reference counting these because they're afraid of the clone. Yes it's nice to have the `AsCalendar` infrastructure, and given that it exists, it's more correct to make `.try_added_with_options()` take `self`, but then we should get rid of the footgun that is a non-`Copy` `AnyCalendar`.

**Comment by robertbastian** [comment]
On a wholly different note:

I think `Date::until` should be `Date::since`. The way you call this method is `a.since(b)`, which is equivalent to `a - b`. `a.until(b)` otoh is equivalent to `b - a`, which is going to lead to confusion and bugs with inverted results.

The Rust standard library types [`std::time::Instant`](https://doc.rust-lang.org/std/time/struct.Instant.html#method.duration_since) and [`std::time::SystemTime`](https://doc.rust-lang.org/std/time/struct.SystemTime.html#method.duration_since) have `::duration_since` methods.

**Comment by Manishearth** [comment]
Actually I think we should just have both. I wanted the initial API to be simple, but we should just have `since` and `until` if that is going to be a problem.

Personally I think both those API names have the problem of confusion; it depends on what you are trying to do.

**Comment by Manishearth** [comment]
> but then we should get rid of the footgun that is a non-Copy AnyCalendar.

I'm in favor of this approach.

Things which are acceptable to me:

 - We make AnyCalendar (and, thus, Date<AC>) Copy so that most people are not forced to think deeply about this.
 - We make it _abundantly clear_ when all of the involved types are cheap clones, so that people do not try to "work around" this
 - We implicitly clone internally

The last option is my least preferred; because the implicit internal clone is counter to the AsCalendar infrastructure.

**Comment by Manishearth** [comment]
> The last option is my least preferred; because the implicit internal clone is counter to the AsCalendar infrastructure.



Given this analysis I opened a PR to remove the implicit clone (#7755). We should pick one of the other two options.

**Comment by Manishearth** [comment]
@markusicu review:

 - try_added_with_options: "returning _a_ new one"
 - Date::try_new_added docs should not mention cloning
 - `largest_unit` is a little confusing, since it may feel like it only returns a single value.
 - We should document the default value of `largest_unit` as Days
 - Discussion around weeks and days in Duration. We could theoretically get rid of it, Markus doesn't have strong opinions. It is nice to be able to specify weeks on the public API.

**Comment by sffc** [comment]
> On a wholly different note:
> 
> I think `Date::until` should be `Date::since`. The way you call this method is `a.since(b)`, which is equivalent to `a - b`. `a.until(b)` otoh is equivalent to `b - a`, which is going to lead to confusion and bugs with inverted results.
> 
> The Rust standard library types [`std::time::Instant`](https://doc.rust-lang.org/std/time/struct.Instant.html#method.duration_since) and [`std::time::SystemTime`](https://doc.rust-lang.org/std/time/struct.SystemTime.html#method.duration_since) have `::duration_since` methods.

I'm open to shipping `since` instead or in addition to `until`. I think I picked `until` because it is the direction specified in Temporal. Awaiting guidance.

**Comment by Manishearth** [comment]
I kind of like until because the usual usage would be `past.until(future)`. But no strong opinion, as long as _internally_ we have it written as until.

We can just expose both. My expectation was that we would be doing that, just not for the MVP.

**Comment by Manishearth** [comment]
#7762 addressed all the issues here.

**Comment by robertbastian** [comment]
`a.until(b)` doesn't seem to return `-(b.until(a))`, why is that? If we had both `until` and `since` I'd want them to only differ by polarity.

**Comment by Manishearth** [comment]
Because it matters what date you start at for the computation. Adding a month to Jan 31 gives Feb 28/29, but subtracting a month from Feb 28 gives Jan 28. `until` tries to add years, then months, then days, to reach the given date. 

Your question actually helps articulate the underlying reason why I like `a.until(b)` as the operation we expose: this is "starting at A, count up to B", unambiguously.

`a.since(b)` on the other hand, would be "starting at b, count up A" by-algorithm, but what people might instead want is "starting at a, count down to B" (`-a.until(b)`).

That's ... somewhat confusing and I would prefer not to have to think about it for 2.2.

**Comment by robertbastian** [comment]
I still think this is surprising behaviour even if just `until` exists. Currently that is documented as just

> Calculating the duration between `other - self`

and a single example where `self` is before `other`. The unintuitive behaviour where `other` is before `self`, which does not produce the inverse of swapping the arguments, is not mentioned. The interaction of theses results with addition is also not mentioned.

**Comment by Manishearth** [comment]
I'm happy to write more docs on this behavior.

**Comment by Manishearth** [comment]
https://github.com/unicode-org/icu4x/pull/7774

**Comment by sffc** [comment]
If we add `since`, it is _defined_ by Temporal in terms of `until` as follows

`date1.since(date2) === date1.until(date2).negated()`

The main reason Temporal has both functions is because if your duration sign is wrong, it's easier to explain switching between since/until than it is to explain why swapping the arguments (`a.until(b)` vs `b.until(a)`) yields different results.

**Comment by Manishearth** [comment]
Yeah, that makes sense. I'm still somewhat inclined to not try to add it for 2.2 but wait until 2.3 to add it, just because we can with minimal cost.
